dbms/src/Operators/LocalSortTransformOp.cpp

```diff
@@ -43,6 +50,82 @@ Block LocalSortTransformOp::getMergeOutput()
     return block;
 }
 
+OperatorStatus LocalSortTransformOp::fromPartialToMerge(Block & block)
+{
+    assert(status == LocalSortStatus::PARTIAL);
+    // convert to merge phase.
+    status = LocalSortStatus::MERGE;
+    if likely (!sorted_blocks.empty())
+    {
+        // In merge phase, the MergeSortingBlocksBlockInputStream of pull model is used to do merge sort.
+        // TODO refine MergeSortingBlocksBlockInputStream and use a more common class to do merge sort in both push model and pull model.
+        merge_impl = std::make_unique<MergeSortingBlocksBlockInputStream>(
+            sorted_blocks,
+            order_desc,
+            log->identifier(),
+            max_block_size,
+            limit);
+        merge_impl->readPrefix();
+        block = getMergeOutput();
+    }
+    return OperatorStatus::HAS_OUTPUT;
+}
+
+OperatorStatus LocalSortTransformOp::fromPartialToRestore()
+{
+    assert(status == LocalSortStatus::PARTIAL);
+    // convert to restore phase.
+    status = LocalSortStatus::RESTORE;
+
+    /// If spill happens
+    LOG_INFO(log, "Begin restore data from disk for merge sort.");
+
+    /// Create sorted streams to merge.
+    spiller->finishSpill();
+    auto inputs_to_merge = spiller->restoreBlocks(0, 0);
+
+    /// Rest of sorted_blocks in memory.
+    if (!sorted_blocks.empty())
+        inputs_to_merge.emplace_back(std::make_shared<MergeSortingBlocksBlockInputStream>(
+            sorted_blocks,
+            order_desc,
+            log->identifier(),
+            max_block_size,
+            limit));
+
+    /// Will merge that sorted streams.
+    merge_impl = std::make_unique<MergingSortedBlockInputStream>(inputs_to_merge, order_desc, max_block_size, limit);
+    merge_impl->readPrefix();
+    return OperatorStatus::IO;
+}
+
+OperatorStatus LocalSortTransformOp::fromPartialToSpill()
+{
+    assert(status == LocalSortStatus::PARTIAL);
+    // convert to restore phase.
+    status = LocalSortStatus::SPILL;
+    assert(!cached_handler);
+    cached_handler = spiller->createCachedSpillHandler(
+        std::make_shared<MergeSortingBlocksBlockInputStream>(sorted_blocks, order_desc, log->identifier(), max_block_size, limit),
+        /*partition_id=*/0,
+        [&]() { return exec_status.isCancelled(); });
+    // fallback to partial phase.
+    if (!cached_handler->batchRead())
+        return fromSpillToPartial();
```

Comment:
How can this happen？

Reply:
It is currently impossible to happen, we can add a `likely` or `RUNTIME_CHECK` here.

Reply:
It's possible, for example, if the query is canceled.